weto/src/main/java/fi/uta/cs/weto/util/NotificationManager.java

```diff
@@ -39,14 +37,225 @@ public void contextInitialized(ServletContextEvent event) {
+                                        try {
+                                            notCompletedSubmissionStudents.remove(sub.getUserId());
+                                        } catch (Exception e) {
+                                            logger.debug("There was no id to remove");
+                                        }
+                                    }
+                                    //In case the permission is for a single user
+                                    if (status != 2 && !isAllUsersPermission) {
+                                        notCompletedSubmissionStudents.add(sub.getUserId());
+                                    }
+                                }
+
+                                for (int student : notCompletedSubmissionStudents) {
+
+
+                                    UserAccount courseAccount2 = UserAccount.select1ById(courseCon,student);
+                                    UserAccount masterAccount2 = UserAccount.select1ByLoginName(masterCon,courseAccount2.getLoginName());
+
+
+                                    int masterCourseID = CourseImplementation.select1ByDatabaseIdAndCourseTaskId(masterCon, databaseID, assignment.getRootTaskId()).getMasterTaskId();
+                                    HashMap<String, String> valueMap = new HashMap<>();
+
+                                    valueMap.put("&task;", assignment.getName());
+                                    valueMap.put("&expirationTime;", new WetoTimeStamp(permission.getEndDate()).toString());
+                                    String message = Notification.getMessageFromTemplate(Notification.SUBMISSION_DEADLINE, valueMap);
+
+                                    // Create link.
+                                    String notificationLink = WetoUtilities.getPackageResource("weto.appBaseUrl")
+                                            + "/viewSubmissions.action?taskId=" + assignment.getId()
+                                            + "&tabId=" + Tab.SUBMISSIONS.getValue()
+                                            + "&dbId=" + databaseID;
+
+                                    Notification notification = new Notification((masterAccount2.getId()), masterCourseID, Notification.SUBMISSION_DEADLINE, notificationLink);
+                                    notification.setMessage(message);
+                                    notification.createNotification(masterCon, courseCon);
+                                }
+                            }
+                        }
+                        connectionManager.freeConnection(courseCon);
+                    }
+                } catch (Exception e) {
+                    logger.error(e);
+                }
+            }
+            connectionManager.freeConnection(masterCon);
+        }
+
+    }
+
+
     private class NotificationEmailTask implements Runnable {

```

Comment:
Undo changes to NotificationEmailTask. Apparently at some point during a merge conflict you've resolved to use the old version that was in your branch instead of the most recent version from master.